nixos/modules/services/networking/firewall.nix

```diff
@@ -510,14 +512,43 @@ in
 
     networking.firewall.extraAllowed = mkOption {
       type = types.listOf (types.submodule { options = {
-           source = mkOption { type = types.string; default = ""; example = "example.com"; description = "Source of the packets"; };
-           protocol = mkOption { type = types.enum ["tcp" "udp" "icmp"]; default = ""; example = "tcp"; description = "Protocol used."; };
-           dport = mkOption { type = types.int; default = 0; example = 22; description = "Destination port of the packet."; };
-           useIPv4 = mkOption { type = types.bool; default = true; description = "Use the IPv4 address of this host."; };
-           useIPv6 = mkOption { type = types.bool; default = true; description = "Use the IPv6 address of this host."; };
+           sourceAddress = mkOption {
+                type = types.nullOr types.string;
+                default = null;
+                example = "198.51.100.0";
+                description = "Source of the packets. According to the iptables manual, DNS host names might be used here, but their use is discouraged (in particular because this delegates the security rule to DNS servers). DNS host names are resolved only when the rule is loaded.";
+           };
+           destAddress = mkOption {
+                type = types.nullOr types.string;
+                default = null;
+                example = "198.51.100.0";
+                description = "Destination of the packets. Same comment as for the `sourceAddress` parameter.";
+           };
+           protocols = mkOption {
+                type = types.nullOr (types.listOf (types.enum ["tcp" "udp" "icmp"]));
```

Comment:
From the manpage of iptables regarding protocols flag:

The protocol of the rule or of the packet to check. The specified protocol can be one of tcp, udp, udplite, icmp, esp, ah, sctp or the special keyword "all", or it can be a numeric value, representing one of these protocols or a different one. A protocol name from /etc/protocols is also allowed. A "!" argument before the protocol inverts the test. The number zero is equivalent to all. "all" will match with all protocols and is taken as default when this option is omitted.